manifest: Close archive once done to umount the device backing the layer

rhvgoyal opened the pull request:
This fixes issue #10593 

Currently newManifest() code can call TarLayer() which returns the archive.
This archive can have some devices mounted (which are backing the layer)
so one needs to call Close() on archive to release those resources once
caller is done with archive.

Looks like newManifest() is not calling Close() on archive and that leads
to some devices not being unmounted when using devicemapper graph drvier.
That also means that if one deletes the image/device later, that deletion
fails as device is still in use (due to previous mount).

So call archive.Close() once archive is not needed anymore to release the
resources backing the archive.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

unclejack commented:
@rhvgoyal Could you add a test for this, please?

rhvgoyal commented:
@unclejack 

This seems to be such low level that I am not sure how to add a test for this. I see that there is already a test for manifest (manifest_test.go) but that test does not catch this problem as they use fakeTar().

In my testing, I faced this problem on older version of docker during "docker push". But new code has removed the call to "manifest" in "docker push" so it does not even seem to trigger. I see that V2 api path calls newManifest() but I don't know how to test that path.

Anyway, adding a test is not a big problem, just that I can't think that exactly what kind of test you are looking for.  If it is about calling archive.close(), then it would not make much sense as it does not
guarantee that actual code does it.

Please give me some idea that what that test is supposed to test.

unclejack commented:
@rhvgoyal Perhaps @dmcgowan and @jlhawn might have some ideas?

dmcgowan commented:
Still investigating this, I seem to be able to reproduce using devicemapper just fine.  Appears that some layers still stay mounted after push though.

rhvgoyal commented:
@dmcgowan 

That's the problem I was facing. If you don't close the archive, then device mapper leaves the device mounted after push. In my test case, I was deleting the local image after the push and while image will be deleted, but actual device and associated metadata deletion will fail as device is still mounted. And that leads to various issues in subequent operations.

rhvgoyal commented:
@dmcgowan 

More details, TarLayer() calls Diff() which in turn gets the device using driver.Get() and later we expect that somebody is going to close archive which in turn will call driver.Put() and unmount the device.

tiborvass commented:
FYI, LGTM bud agree that it'd be better with some tests

dmcgowan commented:
@rhvgoyal it seems this only solves the problem if a manifest generation occurred but no layers were pushed.  When using your code and a push occurred, there were still mounts.  Likely this is because we get the tar layer twice when generating the manifest, one for manifest generation, and again for push.  The same archive closure will need to occur on push after writing the archive to the temp file.

dmcgowan commented:
I also agree this is not something that is easily tested, since I use devicemapper I have a bash command for listing all my docker mounted devices so I can quickly pipe to "xargs umount".

dmcgowan commented:
Adding this fixed the other leftover mounts.  Do you mind adding to your PR?

~~~diff
diff --git a/graph/push.go b/graph/push.go
index 6ae2d67..290c1d0 100644
--- a/graph/push.go
+++ b/graph/push.go
@@ -392,6 +392,8 @@ func (s *TagStore) pushV2Image(r *registry.Session, img *image.Image, endpoint *
        if err != nil {
                return err
        }
+       defer arch.Close()
+
        tf, err := s.graph.newTempFile()
        if err != nil {
                return err
~~~

rhvgoyal commented:
@dmcgowan 

I merged your patch too and rebased my patch on top of current master. But now build failed. Not sure if this is a problem with the patch or it is something unrelated.

dmcgowan commented:
@rhvgoyal thanks for integrating. I will test locally for now, the build failure doesn't appear to be related to your changes.

rhvgoyal commented:
@dmcgowan 

Looks like build succeeded now. I don't know who starts a new build.

Does this pull request look good to you now? If yes, I am wondering who is going to merge it.

dmcgowan commented:
I am debating whether to nit over using multiple Close instead of defer, but either way LGTM

rhvgoyal commented:
Can this PR be merged now?

dmcgowan commented:
Think it needs a few more LGTM from maintainers

ping @jlhawn @tiborvass (for update)

jfrazelle commented on graph/manifest.go:
```diff
@@ -75,12 +75,15 @@ func (s *TagStore) newManifest(localName, remoteName, tag string) ([]byte, error
 
 			tarSum, err := tarsum.NewTarSum(archive, true, tarsum.Version1)
 			if err != nil {
+				archive.Close()
```
cant we defer this?

rhvgoyal replied:
I had thought about using "defer" but I found that it is little tricky here. First of all I can't use a general defer to close it as we are using a for loop inside the function which runs multiple times and archive. Close() needs to happen for each layer.

Now I should be able to use defer in such a way so that it closes archive only in case of error then it can happen that we call close() on archive twice. So say first we do archive.Close() in success path and later failure happens due to some other reason in same function. In that case we will end up calling archive.Close() again. I am not sure if calling Close() on already closed archive is a problem or not.

Is there a way to undefer already scheduled defer?

rhvgoyal replied:
Reading more about defer. Found following.

http://blog.golang.org/defer-panic-and-recover

"A deferred function's arguments are evaluated when the defer statement is evaluated".

That probably means that I can just call "defer archive.Close()" and this can happen in loop and all defer instances should be able to figure out which archive they should be calling close() on when function finishes.

If that works, then only issue remaining will be that all these defer will be called when function returns and devices will remain mounted. Hopefully there is proper reference counting in graph driver backend which can keep count of nested Get() and Put() operations.

rhvgoyal committed:
manifest: Close archive once done using it

manifest code calls TarLayer() and gets archive. This archive needs to
be closed once caller is done using it to release the resrouces held
by archive. For the devicemapper graphdriver, archive keeps a device
mounted (device which is backing the layer). If archive.Close() is not
called, that device remains mounted and later deletion of device fails
leading to various other issues.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
docker-call-close-archive-in-push-v2-image

pushV2Image() calls TarLayer() which returns an archive. One needs to
Close() the archive once done otherwise it will leave mounted devices
if devicemapper graph driver is being used.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> 
Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@jfrazelle 

Ok, pushed a new version of patch which uses defer archive.Close(). Can you please have a look and provide an ack if it looks fine.

jfrazelle commented:
LGTM

jfrazelle commented:
thanks @rhvgoyal
